File: frontend/src/app/models/foia-request.ts
/**
 * FoiaRequest — a Freedom of Information Act request (5 USC 552).
 *
 * Domain reshape from the acquire-gov Solicitation baseline: a FOIA request
 * is intake-driven (requester identity + records sought + fee posture +
 * expedited posture) and governed by a statutory 20-working-day response
 * clock (5 USC 552(a)(6)(A)). The threat model is INVERTED — the requester
 * is an external, potentially adversarial party.
 *
 * NOTE on legacy fields: a handful of acquisition-era optional fields
 * (`naics`, `setAside`, `contractType`, `ceilingValue`, `noticeType`,
 * `sections`, `proposalsDueAt`) remain on the interface. They are the raw
 * material the pair repurposes/deletes across the inherited acquisition
 * components in W4–W5 (per domain-mapping.md §"Additional inherited
 * entities"). They are NOT part of the FOIA shape — do not author new
 * FOIA flows against them.
 */

/** Requester category — drives the FOIA fee category (5 USC 552(a)(4)(A)). */
export type RequesterType =
  | 'commercial'
  | 'news_media_educational_scientific'
  | 'other';

/** Derived fee category per requester type (5 USC 552(a)(4)(A)(ii)). */
export type FeeCategory =
  | 'commercial'                          // search + review + duplication
  | 'news_media_educational_scientific'   // duplication after first 100 pages
  | 'other';                              // search after 2h + dup after 100 pages

/** Final disposition of a FOIA request (DOJ annual-report categories). */
export type FoiaDisposition =
  | 'full_grant'
  | 'partial_grant'
  | 'full_denial'
  | 'no_records';

export interface FoiaRequest {
  id: string;
  /** Owning agency / component (tenant boundary — Item 10 surface). */
  agencyId: string;
  /** Public tracking number assigned at intake (FOIA.gov convention). */
  trackingNumber?: string;
  /** Short subject line for the request. */
  title: string;
  /** Free-text description of the records sought (replaces Section C SOW). */
  recordsSought: string;
  /** Workflow status — see FoiaRequestState. */
  status: string;

  // — Requester identity (inverted threat model: external, untrusted) —
  requesterName?: string;
  requesterOrg?: string;
  requesterType?: RequesterType;

  // — Records scope —
  dateRangeStart?: string;   // ISO date
  dateRangeEnd?: string;     // ISO date

  // — Fee + processing posture —
  /** Derived from requesterType (5 USC 552(a)(4)(A)). */
  feeCategory?: FeeCategory;
  feeWaiverRequested?: boolean;
  expeditedProcessingRequested?: boolean;
  /** Justification supplied when expedited processing is requested. */
  expeditedJustification?: string;

  // — Statutory clock (5 USC 552(a)(6)(A)) —
  /** Date the request was received — starts the 20-working-day clock. */
  receivedDate?: string;
  /** receivedDate + 20 working days (+10 for unusual circumstances). */
  dueDate?: string;

  /** Final disposition once the response is issued. */
  disposition?: FoiaDisposition;

  createdAt?: string;
  updatedAt?: string;

  /**
   * ⚠ LEGACY (acquisition-era) — repurposed/deleted by the pair in W4–W5.
   * Kept optional so the inherited acquisition components still compile.
   */
  description?: string;
  naics?: string;
  setAside?: '' | 'SDVOSB' | 'WOSB' | 'HUBZONE' | '8A' | 'SMALL_BUSINESS' | 'FULL_AND_OPEN';
  contractType?: 'FFP' | 'CPFF' | 'T_AND_M' | 'IDIQ' | 'BPA';
  ceilingValue?: number;
  noticeType?: 'RFI' | 'SOURCES_SOUGHT' | 'RFP' | 'RFQ' | 'COMBINED_SYNOPSIS';
  sections?: FoiaRequestSections;
  proposalsDueAt?: string;
}

/**
 * ⚠ LEGACY (acquisition-era FAR 15.204 Section A–M). Retained only so the
 * inherited acquisition components compile; not part of the FOIA shape.
 */
export interface FoiaRequestSections {
  sectionA?: string;
  sectionB?: string;
  sectionC?: string;
  sectionD?: string;
  sectionE?: string;
  sectionF?: string;
  sectionG?: string;
  sectionH?: string;
  sectionJ?: string;
  sectionK?: string;
  sectionL?: string;
  sectionM?: string;
}

export interface FoiaRequestCreate {
  agencyId: string;
  title: string;
  recordsSought: string;
  status?: string;
  requesterName?: string;
  requesterOrg?: string;
  requesterType?: RequesterType;
  dateRangeStart?: string;
  dateRangeEnd?: string;
  feeCategory?: FeeCategory;
  feeWaiverRequested?: boolean;
  expeditedProcessingRequested?: boolean;
  expeditedJustification?: string;
  receivedDate?: string;
  dueDate?: string;

  // ⚠ LEGACY — see FoiaRequest above.
  description?: string;
  naics?: string;
  setAside?: string;
  contractType?: string;
  ceilingValue?: number;
  noticeType?: string;
  sections?: FoiaRequestSections;
  proposalsDueAt?: string;
}

/**
 * FOIA processing workflow (replaces the solicitation publication lifecycle).
 *   INTAKE_TRIAGE -> EXEMPTION_ANALYSIS -> REDACTION_PROPOSAL -> HITL_REVIEW
 *     -> RESPONSE -> (APPEAL)
 */
export type FoiaRequestState =
  | 'INTAKE_TRIAGE'
  | 'EXEMPTION_ANALYSIS'
  | 'REDACTION_PROPOSAL'
  | 'HITL_REVIEW'
  | 'RESPONSE'
  | 'APPEAL'
  | 'CLOSED';

/**
 * The nine FOIA exemptions (5 USC 552(b)(1)–(b)(9)). Used by RedactionReview.
 */
export type FoiaExemption =
  | '(b)(1)'   // classified national defense / foreign policy
  | '(b)(2)'   // internal personnel rules and practices
  | '(b)(3)'   // prohibited from disclosure by another statute
  | '(b)(4)'   // trade secrets / privileged or confidential commercial info
  | '(b)(5)'   // privileged inter/intra-agency comms (deliberative process)
  | '(b)(6)'   // personal privacy
  | '(b)(7)'   // law-enforcement records
  | '(b)(8)'   // financial-institution supervision
  | '(b)(9)';  // geological / geophysical / well data

export const FOIA_EXEMPTIONS: { code: FoiaExemption; label: string }[] = [
  { code: '(b)(1)', label: '(b)(1) — Classified national defense / foreign policy' },
  { code: '(b)(2)', label: '(b)(2) — Internal personnel rules and practices' },
  { code: '(b)(3)', label: '(b)(3) — Prohibited by other statute' },
  { code: '(b)(4)', label: '(b)(4) — Trade secrets / confidential commercial' },
  { code: '(b)(5)', label: '(b)(5) — Privileged inter/intra-agency (deliberative)' },
  { code: '(b)(6)', label: '(b)(6) — Personal privacy' },
  { code: '(b)(7)', label: '(b)(7) — Law-enforcement records' },
  { code: '(b)(8)', label: '(b)(8) — Financial-institution supervision' },
  { code: '(b)(9)', label: '(b)(9) — Geological / well data' },
];

File: services/foia-request-service/src/main/java/com/karsunfde/foiapipeline/foiarequest/model/FoiaRequest.java
package com.karsunfde.foiapipeline.foiarequest.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

/**
 * FoiaRequest document — a Freedom of Information Act request (5 USC 552).
 *
 * Domain reshape from the acquire-gov Solicitation baseline: the FOIA shape
 * is requester + records-sought + fee posture + expedited posture, governed
 * by a statutory 20-working-day response clock (5 USC 552(a)(6)(A)). The
 * threat model is INVERTED — the requester is an external, potentially
 * adversarial party.
 *
 * ⚠ DELIBERATE — Item 10:
 *   {@code agencyId} is in the schema (so the data is multi-tenant-shaped)
 *   but the repository does not filter on it. Cohort fixes in W2 Wed
 *   multi-tenant retrieval-boundary work.
 *
 * ⚠ DELIBERATE — Item 9:
 *   {@code recordsSought} / {@code description} are not sanitized; arbitrary
 *   HTML accepted on write and returned verbatim on read. Cohort fixes in
 *   W4 Wed AI Security Engineering Day (prompt-injection-via-stored-content
 *   — the records-sought text feeds the ai-orchestrator prompt; the inverted
 *   threat model makes this especially live).
 *
 * State machine (FOIA processing):
 *   INTAKE_TRIAGE -> EXEMPTION_ANALYSIS -> REDACTION_PROPOSAL -> HITL_REVIEW
 *     -> RESPONSE -> (APPEAL) -> CLOSED
 *
 * NOTE: the legacy acquisition fields ({@code naics}, {@code setAside},
 * {@code sections}, {@code postedAt}, {@code closingAt}) and their accessors
 * remain so the inherited acquisition components/controllers compile. They
 * are NOT part of the FOIA shape; the pair repurposes/removes them W4–W5.
 */
@Document(collection = "foia_requests")
public class FoiaRequest {

    @Id
    private String id;

    /** ⚠ Item 10 — present but un-enforced. */
    private String agencyId;

    /** Public tracking number assigned at intake (FOIA.gov convention). */
    private String trackingNumber;

    private String title;

    /** Free-text description of the records sought. ⚠ Item 9 — accepts arbitrary HTML. */
    private String recordsSought;

    private String status;

    // --- Requester (inverted threat model: external, untrusted) ---
    private String requesterName;
    private String requesterOrg;
    /** commercial | news_media_educational_scientific | other. */
    private String requesterType;

    // --- Records scope ---
    private Instant dateRangeStart;
    private Instant dateRangeEnd;

    // --- Fee + processing posture ---
    /** Derived from requesterType (5 USC 552(a)(4)(A)). */
    private String feeCategory;
    private boolean feeWaiverRequested;
    private boolean expeditedProcessingRequested;
    private String expeditedJustification;

    // --- Statutory clock (5 USC 552(a)(6)(A)) ---
    /** Starts the 20-working-day clock. */
    private Instant receivedDate;
    /** receivedDate + 20 working days (+10 unusual circumstances). */
    private Instant dueDate;

    /** full_grant | partial_grant | full_denial | no_records. */
    private String disposition;

    /** ⚠ LEGACY (acquisition-era) — kept un-sanitized to preserve Item 9. */
    private String description;
    /** ⚠ LEGACY. */
    private String naics;
    /** ⚠ LEGACY. */
    private String setAside;
    /** ⚠ LEGACY — values unsanitized (Item 9). */
    private Map<String, String> sections = new HashMap<>();
    /** ⚠ LEGACY. */
    private Instant postedAt;
    /** ⚠ LEGACY. */
    private Instant closingAt;

    private Instant createdAt;
    private Instant updatedAt;

    public FoiaRequest() {}

    // --- getters / setters ---

    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getAgencyId() { return agencyId; }
    public void setAgencyId(String agencyId) { this.agencyId = agencyId; }

    public String getTrackingNumber() { return trackingNumber; }
    public void setTrackingNumber(String trackingNumber) { this.trackingNumber = trackingNumber; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getRecordsSought() { return recordsSought; }
    public void setRecordsSought(String recordsSought) { this.recordsSought = recordsSought; }

    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }

    public String getRequesterName() { return requesterName; }
    public void setRequesterName(String requesterName) { this.requesterName = requesterName; }

    public String getRequesterOrg() { return requesterOrg; }
    public void setRequesterOrg(String requesterOrg) { this.requesterOrg = requesterOrg; }

    public String getRequesterType() { return requesterType; }
    public void setRequesterType(String requesterType) { this.requesterType = requesterType; }

    public Instant getDateRangeStart() { return dateRangeStart; }
    public void setDateRangeStart(Instant dateRangeStart) { this.dateRangeStart = dateRangeStart; }

    public Instant getDateRangeEnd() { return dateRangeEnd; }
    public void setDateRangeEnd(Instant dateRangeEnd) { this.dateRangeEnd = dateRangeEnd; }

    public String getFeeCategory() { return feeCategory; }
    public void setFeeCategory(String feeCategory) { this.feeCategory = feeCategory; }

    public boolean isFeeWaiverRequested() { return feeWaiverRequested; }
    public void setFeeWaiverRequested(boolean feeWaiverRequested) { this.feeWaiverRequested = feeWaiverRequested; }

    public boolean isExpeditedProcessingRequested() { return expeditedProcessingRequested; }
    public void setExpeditedProcessingRequested(boolean expeditedProcessingRequested) { this.expeditedProcessingRequested = expeditedProcessingRequested; }

    public String getExpeditedJustification() { return expeditedJustification; }
    public void setExpeditedJustification(String expeditedJustification) { this.expeditedJustification = expeditedJustification; }

    public Instant getReceivedDate() { return receivedDate; }
    public void setReceivedDate(Instant receivedDate) { this.receivedDate = receivedDate; }

    public Instant getDueDate() { return dueDate; }
    public void setDueDate(Instant dueDate) { this.dueDate = dueDate; }

    public String getDisposition() { return disposition; }
    public void setDisposition(String disposition) { this.disposition = disposition; }

    // --- LEGACY accessors (inherited acquisition components) ---

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public String getNaics() { return naics; }
    public void setNaics(String naics) { this.naics = naics; }

    public String getSetAside() { return setAside; }
    public void setSetAside(String setAside) { this.setAside = setAside; }

    public Map<String, String> getSections() { return sections; }
    public void setSections(Map<String, String> sections) { this.sections = sections; }

    public Instant getPostedAt() { return postedAt; }
    public void setPostedAt(Instant postedAt) { this.postedAt = postedAt; }

    public Instant getClosingAt() { return closingAt; }
    public void setClosingAt(Instant closingAt) { this.closingAt = closingAt; }

    public Instant getCreatedAt() { return createdAt; }
    public void setCreatedAt(Instant createdAt) { this.createdAt = createdAt; }

    public Instant getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(Instant updatedAt) { this.updatedAt = updatedAt; }
}
